File: C++/Leetcode/LC22-GenerateParentheses.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    // Recursive helper function to keep track of recursion tree so far and either append to ans or continue
    void solve(int leftRem, int rightRem, std::string currStr, std::vector<std::string>& ans) {
        if (leftRem == 0 && rightRem == 0) {
            ans.push_back(currStr);
            return;
        }
        if (leftRem == rightRem || rightRem == 0) {
            solve(leftRem-1, rightRem, currStr+"(", ans);
        } else if (leftRem == 0) {
            solve(leftRem, rightRem-1, currStr+")", ans);
        } else {
            // We have two different options in this case, and we need to cover both of them
            solve(leftRem-1, rightRem, currStr+"(", ans);
            solve(leftRem, rightRem-1, currStr+")", ans);
        }
    }

    std::vector<std::string> generateParenthesis(int n) {
        std::vector<std::string> ans; 
        solve(n, n, "", ans); // solve will operate on ans and populate it with all valid combinations
        return ans;
    }
};

File: go/Leetcode/Stack/LC22-GenerateParentheses.go
package leetcode

func generateParenthesis(n int) []string {
	var ans []string
	solve(n, n, "", &ans)
	return ans
}

// Helper function keeping track of recusion tree so far, appending parentheses based on case
func solve(l int, r int, curr string, ans *[]string) {
	if l == 0 && r == 0 {
		*ans = append(*ans, curr)
		return
	}

	if l == r || r == 0 {
		solve(l-1, r, curr+"(", ans)
	} else if l == 0 {
		solve(l, r-1, curr+")", ans)
	} else {
		// Try both remaining options
		solve(l-1, r, curr+"(", ans)
		solve(l, r-1, curr+")", ans)
	}
}
